src/index.ts

```diff
@@ -854,6 +870,63 @@ export abstract class PinejsClientCore<PinejsClient> {
 		}
 	}
 
+	private canRetryDefaultHandler(err: any) {
+		const code = err?.statusCode;
+		return code == null || code === 429 || (code >= 500 && code < 600);
+	}
+
+	protected async callWithRetry<T>(
+		fnCall: () => Promise<T>,
+		retry: RetryParameters,
+	): Promise<T> {
+		if (!retry || !this.retry) {
+			return await fnCall();
+		}
+
+		const minDelayMs = retry.minDelayMs ?? this.retry.minDelayMs ?? 0;
+		const maxDelayMs = retry.maxDelayMs ?? this.retry.maxDelayMs ?? 0;
+		const maxAttempts = retry.maxAttempts ?? this.retry.maxAttempts ?? 0;
+
+		if (minDelayMs <= 0) {
+			throw new Error('pinejs-client minDelayMs must be positive');
+		}
+		if (maxDelayMs <= 0) {
+			throw new Error('pinejs-client maxDelayMs must be positive');
+		}
+		if (maxAttempts <= 0) {
+			throw new Error('pinejs-client maxAttempts be positive');
+		}
+		if (minDelayMs > maxDelayMs) {
+			throw new Error(
+				'pinejs-client maxDelayMs must be greater than or equal to minDelayMs',
+			);
+		}
+
+		const onRetryHandler = retry.onRetry ?? this.retry.onRetry;
+
+		let attempt = 1;
+
+		const canRetryHandler =
+			retry.canRetry ?? this.retry.canRetry ?? this.canRetryDefaultHandler;
+
+		while (true) {
+			try {
+				return await fnCall();
+			} catch (err) {
+				if (attempt >= maxAttempts || !canRetryHandler(err)) {
+					throw err;
+				}
+
+				const delayMs = Math.min(2 ** (attempt - 1) * minDelayMs, maxDelayMs);
+				attempt++;
+				onRetryHandler?.(err, delayMs, attempt, maxAttempts);
```

Comment:
note that `attempt` is incremented _before_ calling `onRetryHandler` because retries effectively begin with attempt number 2.

Reply:
I'd add this as a comment in the code itself